src/Sulu/Bundle/MediaBundle/EventListener/MediaSecurityListener.php

```diff
@@ -0,0 +1,84 @@
+<?php
+
+/*
+ * This file is part of Sulu.
+ *
+ * (c) Sulu GmbH
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Bundle\MediaBundle\EventListener;
+
+use Sulu\Bundle\MediaBundle\Admin\MediaAdmin;
+use Sulu\Bundle\MediaBundle\Entity\Collection;
+use Sulu\Bundle\MediaBundle\Entity\CollectionRepositoryInterface;
+use Sulu\Component\Security\Authorization\PermissionTypes;
+use Sulu\Component\Security\Authorization\SecurityCheckerInterface;
+use Sulu\Component\Security\Authorization\SecurityCondition;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\HttpKernel\Event\RequestEvent;
+use Symfony\Component\HttpKernel\Exception\NotFoundHttpException;
+
+/**
+ * @internal this class is for internal usages
+ */
+class MediaSecurityListener implements EventSubscriberInterface
+{
+    /**
+     * @var CollectionRepositoryInterface
+     */
+    private $collectionRepository;
+
+    /**
+     * @var SecurityCheckerInterface
+     */
+    private $securityChecker;
+
+    public function __construct(
+        CollectionRepositoryInterface $collectionRepository,
+        SecurityCheckerInterface $securityChecker
+    ) {
+        $this->collectionRepository = $collectionRepository;
+        $this->securityChecker = $securityChecker;
+    }
+
+    public static function getSubscribedEvents(): array
+    {
+        return [
+            RequestEvent::class => 'onKernelRequest',
+        ];
+    }
+
+    public function onKernelRequest(RequestEvent $event): void
+    {
+        $request = $event->getRequest();
+
+        if (!$event->isMasterRequest()) {
+            return;
+        }
+
+        $routeName = $request->attributes->get('_route');
+
+        if (!\in_array($routeName, ['sulu_media.website.media.download', 'sulu_media.website.media.download_admin'])) {
```

Comment:
Isn't that code called now for *every request*? And both of these routes are referring to the same controller, right? Wouldn't it be better performance-wise to do that in the controller instead?

Reply:
If somebody did override the route/controller, the security check would then not longer work.

I thought about using attributes instead maybe so I could check for `$request->attributes->getBoolean('media_security');` so I can enable the security check on the route definition a default attribute. What do you think?

Reply:
I think moving this check to a separate listener is very confusing, when it could also be attached to a single controller. And in addition to that it has a negative impact on the entire system, because this check has to be done for every single route. For these reasons I would not do it this way, but maybe add it to the `UPGRADE.md` instead, so that people know they have to add the security check.

Using attributes instead does not improve performance, does it? And it introduces even more indirection.

PS: I think we should make a lot more stuff `final` :see_no_evil: Would even make it the new default.

Reply:
As discussed in the daily I will refractor it into the controller